File: plugins/collector/event/cron.go
package event

import (
	"bufio"
	"collector/eventmanager"
	"collector/utils"
	"crypto/md5"
	"errors"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/chriskaliX/SDK"
	"github.com/chriskaliX/SDK/transport/protocol"
	lru "github.com/hashicorp/golang-lru"
	"github.com/mitchellh/mapstructure"
)

var (
	CronSearchDirs = []string{
		"/etc/cron.d",
		"/var/spool/cron/",
		"/var/spool/cron/crontabs",
	}
)

const CRON_DATATYPE = 2001

var _ eventmanager.IEvent = (*Crontab)(nil)

type Cron struct {
	Minute     string `mapstructure:"minute"`
	Hour       string `mapstructure:"hour"`
	DayOfMonth string `mapstructure:"day_of_month"`
	Month      string `mapstructure:"month"`
	DayOfWeek  string `mapstructure:"day_of_week"`
	User       string `mapstructure:"user"`
	Command    string `mapstructure:"command"`
	Path       string `mapstructure:"path"`
}

type Crontab struct {
	cronCache *lru.Cache
}

func (Crontab) DataType() int { return CRON_DATATYPE }

func (Crontab) DataTypeSync() int { return 3001 }

func (n *Crontab) Flag() eventmanager.EventMode { return eventmanager.Periodic }

func (Crontab) Name() string { return "cron" }

func (Crontab) Immediately() bool { return false }

// https://github.com/osquery/osquery/blob/d2be385d71f401c85872f00d479df8f499164c5a/tests/integration/tables/crontab.cpp
// const std::string kSystemCron = "/etc/crontab";

// const std::vector<std::string> kCronSearchDirs = {
// 	"/etc/cron.d/", // system all
// 	"/var/at/tabs/", // user mac:lion
// 	"/var/spool/cron/", // user linux:centos
// 	"/var/spool/cron/crontabs/", // user linux:debian
// };
// https://github.com/osquery/osquery/blob/2c2b85cbd25a381eb0973017427928e5691c4431/osquery/tables/system/posix/crontab.cpp
func (c *Crontab) parse(withUser bool, path string, file *os.File) (crons []Cron) {
	r := bufio.NewScanner(io.LimitReader(file, 1024*1024))
	for r.Scan() {
		line := r.Text()
		// skip empty and null line
		if line != "" && strings.TrimSpace(line)[0] == '#' {
			continue
		} else if strings.Contains(line, "@reboot") {
			fields := strings.Fields(line)
			cron := Cron{
				Minute:     "@reboot",
				Hour:       "@reboot",
				DayOfMonth: "@reboot",
				Month:      "@reboot",
				DayOfWeek:  "@reboot",
				Path:       path,
			}
			if len(fields) >= 2 {
				if withUser {
					cron.User = file.Name()
					cron.Command = strings.Join(fields[1:], " ")
				} else if len(fields) >= 3 {
					cron.User = fields[1]
					cron.Command = strings.Join(fields[2:], " ")
				}
			}
			crons = append(crons, cron)
		} else {
			fields := strings.Fields(line)
			if len(fields) >= 6 {
				cron := Cron{
					Minute:     fields[0],
					Hour:       fields[1],
					DayOfMonth: fields[2],
					Month:      fields[3],
					DayOfWeek:  fields[4],
					Path:       path,
				}
				if withUser {
					cron.User = filepath.Base(file.Name())
					cron.Command = strings.Join(fields[5:], " ")
				} else if len(fields) >= 7 {
					cron.User = fields[5]
					cron.Command = strings.Join(fields[6:], " ")
				}
				crons = append(crons, cron)
			}
		}
	}
	return
}

func (c *Crontab) GetCron() (crons []Cron, err error) {
	for _, dir := range CronSearchDirs {
		err := filepath.Walk(dir, func(path string, info fs.FileInfo, err error) error {
			if err != nil {
				return err
			}
			if info.Mode().IsRegular() {
				f, err := os.Open(path)
				if err != nil {
					return nil
				}
				flag := strings.HasPrefix(path, "/var/spool/cron")
				crons = append(crons, c.parse(flag, path, f)...)
				f.Close()
			}
			return nil
		})
		if err != nil {
			continue
		}
	}

	if f, e := os.Open("/etc/crontab"); e == nil {
		crons = append(crons, c.parse(false, "/etc/crontab", f)...)
		f.Close()
	}

	if len(crons) == 0 {
		err = errors.New("crontab is empty")
	}
	return
}

func (c *Crontab) Run(s SDK.ISandbox, sig chan struct{}) (err error) {
	if c.cronCache == nil {
		c.cronCache, _ = lru.New(240)
	}
	hash := utils.Hash()
	// get crons if it start
	if crons, err := c.GetCron(); err == nil {
		for _, cron := range crons {
			c.cronCache.Add(md5.Sum([]byte(cron.Command)), true)
		}
		for _, cron := range crons {
			rec := &protocol.Record{
				DataType:  int32(c.DataType()),
				Timestamp: time.Now().Unix(),
				Data: &protocol.Payload{
					Fields: make(map[string]string, 9),
				},
			}
			mapstructure.Decode(&cron, &rec.Data.Fields)
			rec.Data.Fields["package_seq"] = hash
			s.SendRecord(rec)
		}
	}
	return nil
}

File: plugins/collector/src/event/cron.rs
/// Cron collector — data_type 2001.
/// Mirrors Go's event/cron.go.  Reusable parse() function is also consumed by
/// cron_watcher.rs.
use std::collections::HashMap;
use std::fs;
use std::io::{BufRead, BufReader};
use std::time::Duration;

use anyhow::Result;
use async_trait::async_trait;
use sdk::Client;

use crate::manager::{EventMode, IEvent};
use crate::event::{hash, make_record};

const DATA_TYPE: i32 = 2001;

pub static CRON_DIRS: &[&str] = &[
    "/etc/cron.d",
    "/var/spool/cron",
    "/var/spool/cron/crontabs",
];
pub const ETC_CRONTAB: &str = "/etc/crontab";

pub struct Cron;

#[async_trait]
impl IEvent for Cron {
    fn name(&self)        -> &'static str { "cron" }
    fn data_type(&self)   -> i32          { DATA_TYPE }
    fn flag(&self)        -> EventMode    { EventMode::Periodic }
    fn immediately(&self) -> bool         { false }

    async fn run(&mut self, client: &mut Client) -> Result<()> {
        let seq = hash();
        // Collect all cron files
        let mut files: Vec<(String, bool)> = Vec::new(); // (path, with_user)

        // /etc/crontab — has user field
        files.push((ETC_CRONTAB.to_owned(), true));

        for dir in CRON_DIRS {
            let with_user = dir.starts_with("/var/spool/cron");
            if let Ok(rd) = fs::read_dir(dir) {
                for entry in rd.flatten() {
                    let path = entry.path().to_string_lossy().to_string();
                    files.push((path, with_user));
                }
            }
        }

        for (path, with_user) in files {
            let Ok(content) = fs::read_to_string(&path) else { continue };
            let entries = parse(with_user, &path, &content);
            for mut fields in entries {
                fields.insert("package_seq".into(), seq.clone());
                let _ = client.send_record(&make_record(DATA_TYPE, fields.clone()));
                tokio::time::sleep(Duration::from_millis(5)).await;
            }
        }
        Ok(())
    }
}

// ── Public parse helper (shared with cron_watcher) ──────────────────────────

pub struct CronEntry {
    pub fields: HashMap<String, String>,
}

/// Parse a crontab file.  `with_user=true` means the 6th field is the username
/// (as in /etc/crontab and /etc/cron.d/ files).
pub fn parse(with_user: bool, path: &str, content: &str) -> Vec<HashMap<String, String>> {
    let mut result = Vec::new();
    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') { continue; }

        let mut m: HashMap<String, String> = HashMap::new();
        m.insert("path".into(), path.to_owned());

        if let Some(rest) = line.strip_prefix('@') {
            // @reboot / @hourly / etc.
            let (keyword, command) = rest.split_once(char::is_whitespace)
                .unwrap_or((rest, ""));
            m.insert("minute".into(),       format!("@{keyword}"));
            m.insert("hour".into(),         String::new());
            m.insert("day_of_month".into(), String::new());
            m.insert("month".into(),        String::new());
            m.insert("day_of_week".into(),  String::new());
            if with_user {
                let (user, cmd) = command.split_once(char::is_whitespace).unwrap_or(("", command));
                m.insert("user".into(),    user.to_owned());
                m.insert("command".into(), cmd.to_owned());
            } else {
                m.insert("user".into(),    String::new());
                m.insert("command".into(), command.to_owned());
            }
            result.push(m);
            continue;
        }

        let cols: Vec<&str> = line.splitn(7, char::is_whitespace)
            .filter(|s| !s.is_empty())
            .collect();

        if with_user && cols.len() >= 7 {
            m.insert("minute".into(),       cols[0].to_owned());
            m.insert("hour".into(),         cols[1].to_owned());
            m.insert("day_of_month".into(), cols[2].to_owned());
            m.insert("month".into(),        cols[3].to_owned());
            m.insert("day_of_week".into(),  cols[4].to_owned());
            m.insert("user".into(),         cols[5].to_owned());
            m.insert("command".into(),      cols[6..].join(" "));
        } else if !with_user && cols.len() >= 6 {
            m.insert("minute".into(),       cols[0].to_owned());
            m.insert("hour".into(),         cols[1].to_owned());
            m.insert("day_of_month".into(), cols[2].to_owned());
            m.insert("month".into(),        cols[3].to_owned());
            m.insert("day_of_week".into(),  cols[4].to_owned());
            m.insert("user".into(),         String::new());
            m.insert("command".into(),      cols[5..].join(" "));
        } else {
            continue;
        }
        result.push(m);
    }
    result
}
